rfc/0018-certificate-handling.md

```diff
@@ -0,0 +1,259 @@
+dependency against CertManager. The new solution should also take into account
+certificate rotation.
+
+## Examples / User Stories
+
+> As a Kubernetes operator, I don't want to install CertManager
+> in order to deploy Kubearden.
+
+> As a Kubernetes operator,
+> I want Kubewarden certificates to be automatically rotated
+> before they reach their expiration date.
+
+# Detailed design
+
+## Components that make use of TLS certificates
+
+All the certificates used by the Kubewarden stack are internal to the cluster.
+All of them are used to secure the communication between the API Server and different
+HTTPS endpoints.
+
+These are the HTTP servers managed by Kubewarden that require TLS termination:
+
+- kubewarden-controller: this is a Validating and Mutating webhook endpoint that
+  processes the Kubewarden CRDs (like `PolicyServer`, `ClusterAdmissionPolicy`
+  and `AdmissionPolicy`)
+- Policy Server: each Deployment has its own dedicated certificate
+
+> **Note:** all these endpoints are either Validating or Mutating webhook
+> endpoints for Kubernetes
+
+Currently, Kubewarden 1.6 at the time of wriging, the certificate used by the
+kubewarden-controller is generated by CertManager.
+On the other hand, the certificates used by the
+Policy Server instances are generated by our controller.
+
+The kubewarden-controller generates a self-signed Certificate Authority (CA) that
+is then used to sign all the Policy Server certificates.
+
+## Webhook Configuration and Certificates
+
+Both `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration`
+resources have a
+[`clientConfig`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#webhookclientconfig-v1-admissionregistration-k8s-io)
+object inside of their definition.
+The `clientConfig.caBundle` has the bundle of CA required to
+verify the TLS termination of the webhook endpoint:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
+webhooks:
+  - name: my-webhook.example.com
+    clientConfig:
+      caBundle: <CA_BUNDLE>
+      service:
+        namespace: my-service-namespace
+        name: my-service-name
+        path: /my-path
+        port: 1234
+```
+
+We have one `MutatingWebhookConfiguration` resource used by the kubewarden-controller which
+holds a reference to the CA managed by CertManager.
+
+Each `ClusterAdmissionPolicy` and `AdmissionPolicy` has a dedicated `(Validating|Mutating)WebhookConfiguration`
+object defined. For all of them, the `clientConfig.caBundle` contains the certificate of
+the CA generated by our controller.
+
+## Proposed Design
+
+We would like to get rid of CertManager. To do that we need to change how certificates are
+managed for the `kubewarden-controller`.
+Going forward, the controller will also take care of generating the certificate used by the kubewarden-controller.
+This certificate is going to be signed by the CA which is already created by the controller.
```

Comment:
Will we remove it completely or make it optional? I would like to double check that because the original issue around this topic says to turn it [optional](https://github.com/kubewarden/kubewarden-controller/issues/422#issuecomment-1644206549)

Reply:
I agree on having it optional, and disabled by default.

Reply:
+1

Reply:
@flavio does this comment still stand? I thought we wanted to still support cert-manager, but not as default.

Reply:
Last week @jvanz, @fabriziosestito and agreed to focus on removing the cert-manager dependency. We can make it optional in the future if there's some request coming from the community.